SpatialGDK/Source/SpatialGDK/Private/Interop/CrossServerRPCHandler.cpp

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/CrossServerRPCHandler.h"
+
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "SpatialGDKSettings.h"
+
+DEFINE_LOG_CATEGORY(LogCrossServerRPCHandler);
+
+namespace SpatialGDK
+{
+CrossServerRPCHandler::CrossServerRPCHandler(ViewCoordinator& InCoordinator, TUniquePtr<RPCExecutorInterface> InRPCExecutor)
+	: Coordinator(InCoordinator)
+	, RPCExecutor(MoveTemp(InRPCExecutor))
+{
+}
+
+void CrossServerRPCHandler::ProcessOps(const TArray<Worker_Op>& WorkerMessages)
+{
+	for (const auto& Op : WorkerMessages)
+	{
+		if (Op.op_type == WORKER_OP_TYPE_COMMAND_REQUEST
+			&& Op.op.command_request.request.command_index == SpatialConstants::UNREAL_RPC_ENDPOINT_COMMAND_ID
+			&& Op.op.command_request.request.component_id == SpatialConstants::SERVER_TO_SERVER_COMMAND_ENDPOINT_COMPONENT_ID)
+		{
+			HandleWorkerOp(Op);
+		}
+	}
+
+	ProcessPendingCrossServerRPCs();
+
+	while (RPCsToDelete.Num() > 0 && RPCsToDelete.HeapTop().Key < FDateTime::Now())
+	{
+		RPCsToDelete.HeapPopDiscard();
+	}
+}
+
+void CrossServerRPCHandler::ProcessPendingCrossServerRPCs()
+{
+	for (auto CrossServerRPCs = QueuedCrossServerRPCs.CreateIterator(); CrossServerRPCs; ++CrossServerRPCs)
+	{
+		int32 ProcessedRPCs = 0;
+		for (const auto& Command : CrossServerRPCs.Value())
+		{
+			if (!TryExecuteCrossServerRPC(Command))
+			{
+				break;
+			}
+
+			RPCGuidsInFlight.Remove(Command.Payload.UniqueId);
+			RPCsToDelete.HeapPush(TTuple<FDateTime, uint32>(FDateTime::Now() + Command.TimeoutMillis, Command.Payload.UniqueId));
+			++ProcessedRPCs;
+		}
+
+		CrossServerRPCs.Value().RemoveAt(0, ProcessedRPCs);
+		if (CrossServerRPCs.Value().Num() == 0)
+		{
+			CrossServerRPCs.RemoveCurrent();
+		}
+	}
+}
+
+const TMap<Worker_EntityId_Key, TArray<FCrossServerRPCParams>>& CrossServerRPCHandler::GetQueuedCrossServerRPCs() const
+{
+	return QueuedCrossServerRPCs;
+}
+
+void CrossServerRPCHandler::HandleWorkerOp(const Worker_Op& Op)
+{
+	const Worker_CommandRequestOp& CommandOp = Op.op.command_request;
+	FCrossServerRPCParams Params = RPCExecutor->TryRetrieveCrossServerRPCParams(Op);
+
+	if (Params.RequestId == -1)
```

Comment:
- if it can't find the function
- if the payload is badly formatted (schema mismatch)
- if it can't retrieve the target uobject
- if the entity is still waiting for async load
- if the RPC is for some reason actually not a cross server rpc